### ISSUE TITLE
Add tone mapping for HDR to SDR transcoding

### ISSUE THREAD

**Issue by LukePulverenti** [MediaBrowser/Emby#3480]

**Comment by preepe** [comment]
Great to hear! I am looking forward to this feature!

**Comment by tchirou** [comment]
Follwing this
Thanks !

**Comment by Lord-Malgus** [comment]
Any updates / news on this issue?   Thank you.

**Comment by softworkz** [comment]
What makes this a bit complicated is the integration of this in combination with hardware acceleration. Ideally, the tone mapping would be done in hardware as well.

Just this week, a VAAPI hw tonemapping filter has been added to ffmpeg and I heard that another one for QuickSync might be added soon. For Nvidia, I'd know how to implement it (roughly), but hopefully somebody else will do it soon.

Before we can go into this direction, we got some base work to do on our side, but that's all planned and will happen soon.

**Comment by Lord-Malgus** [comment]
Really appreciate the update.  Thank you sir!

**Comment by ianken** [comment]
Thanks to the dev folks for responding. The other media server solution guys are basically ignoring this.

**Comment by gohlas** [comment]
Very welcome feature! That would be great.

**Comment by jaroslawjanas** [comment]
@softworkz 
Can you define "soon" please ? 

Jellyfin seems to be ahead of you guys:
- [nvidia support](https://github.com/jellyfin/jellyfin/pull/3442)
- [amd support](https://github.com/jellyfin/jellyfin/pull/4171).

Thanks.

**Comment by LukePulverenti** [comment]
> @softworkz
> Can you define "soon" please ?
> 
> Jellyfin seems to be ahead of you guys:
> 
> * [nvidia support](https://github.com/jellyfin/jellyfin/pull/3442)
> * [amd support](https://github.com/jellyfin/jellyfin/pull/4171).
> 
> Thanks.

Don't have an ETA but it's moving up the list.

**Comment by softworkz** [comment]
> @softworkz
> Can you define "soon" please ?
> 
> Jellyfin seems to be ahead of you guys:
> 
> * [nvidia support](https://github.com/jellyfin/jellyfin/pull/3442)
> * [amd support](https://github.com/jellyfin/jellyfin/pull/4171).
> 
> Thanks.

Those PRs don't make much sense to me. I totally doubt that it's working with AMD/AMF and there seem to be a lot of limitations in case of Nvidia.
We don't work like this. When we deliver a feature like that, it needs to be rock-solid - not only for cases where it might be working, but also detecting the other cases up-front, in order to keep hardware acceleration working fine in all non-supported scenarios.

**Have you actually tested this and got it working successfully on your side?**

...or are you just reading their pull request history?

**Comment by jaroslawjanas** [comment]
@softworkz 
I am by no means comparing your work process to jellyfin's, I put those two links to PRs as I thought It might be of interest to you.

I haven't tested jellyfin's implementation yet but I do have it in my schedule.

I found those jellyfin PRs as I was researching the same issue in Emby and since there has been no updates for a while, and the ETA is always "soon", I asked you for an update.

I am sure, you are doing your best guys.
Keep up the good work ❤️.

**Comment by softworkz** [comment]
It's a complicated subject unfortunately. There has been another PR at JF a while ago about sw tonemapping (without hw acceleration) and only when you read everything to the end in all detail, you'll be able to realize that there is absolutely no value in doing this. 
Why? - Because it runs like about 100 times slower than real-time on a mid-range system, without even having a chance to run in realtime on any high-end system.

I don't think that the required amount of computing is really that high, but that's the only way that ffmpeg provides at this time.
In the other threads, you have probably read about all the limitations: like that vaapi tonemapping works with the very latest cpu models only, and that there is no tonemapping with Nvidia directly, only via OpenCL, which is an additional hwa framework bringing its own limitations and issues into the game. Again - not suitable for the masses.

I have - meanwhile successfully - tried to trigger some progress at the side of Intel QuickSync (https://github.com/Intel-Media-SDK/MediaSDK/issues/1949#issuecomment-720564734), but that will need to be integrated into ffmpeg first, and eventually it might have the same limitations like the vaapi-tonemap feature (both are accessing the same core hw feaures), so we'll have to see what happens there.

I'm constantly monitoring progress at the side of ffmpeg and at the side of the hardware vendors. As soon as the situation changes in a way that we can come up with something that would realistically work for a relevant amount of Emby users, we'll be ready to go for it.

But as long as this remains to be something that would work at most for 3% or 5% of all Emby users and then only under certain conditions, it wouldn't be worth the effort and even more a sad (and time-intensive) story having to explain all the other 97% why it doesn't work for them.

**Comment by softworkz** [comment]
One more note. Just look at the first comparison image they are presenting there:

![image](https://user-images.githubusercontent.com/4985349/99549330-d12ce100-29b9-11eb-9646-2345516a1e4c.png)

At first sight, the right side surely appears to be great with lots of intense colors compared to the left.
But now, zoom into the right picture and magnify it to fill the screen. Then answer the following question:

- Does reality look like that?
  => No, it doesn't

HDR is not about delivering over-saturated colors. It's about delivering a more realistic picture. And that picture above is far away from realistic (just look at the windows of the car and the sky). And far away from its intended way of presentation.

- Download the original Sony Demo Video
- View it on your HDR TV
  => you will see that the original video looks totally different than this right-side comparison picture

Whatever they've done there - it's totally wrong. 
This is in no way the result of an accurate tone-mapping algorithm.

**Comment by ryester19** [comment]
Accurate or not, I would still prefer the right image over the left when I'm forced to watch HDR content on a SDR screen. It just needs to be made known to end users that tonemapping isn't a replacement to buying a HDR screen, and that you shouldn't bug the developers if things appear wrong to you. Offering every possible customization option might limit complaints?

**Comment by cybergrimes** [comment]
Confused by that 4 year old demo file... MediaInfo reports color primaries BT 709 with a bit depth of 8 bits.
Shouldn't these be BT 2020 and 10 bit for an HDR10 encoded video?

**Comment by softworkz** [comment]
> Confused by that 4 year old demo file... MediaInfo reports color primaries BT 709 with a bit depth of 8 bits.
> Shouldn't these be BT 2020 and 10 bit for an HDR10 encoded video?

Get it from here: https://drive.google.com/drive/folders/19x0YQoaM4tFI5zYTLKnoHMcGIpmRSR8E

**Comment by softworkz** [comment]
> Accurate or not, I would still prefer the right image over the left when I'm forced to watch HDR content on a SDR screen. It just needs to be made known to end users that tonemapping isn't a replacement to buying a HDR screen

Let me put it like this: tone-mapping is done right, when an HDR video that you have will look just the same like all your other SDR videos (on your SDR screen). As simple as that and surely not with overdriven contrast and saturation.

I should add the following for completeness (like I had stated a number of times in the Emby forums): 

Even when tonemapping is done right, and the result would basically look similar to regular SDR videos, this doesn't mean that the result would be equal to the originally released SDR version, where the luminance and chrominance ranges have been manually optimized for each individual scene.

**Comment by darthShadow** [comment]
Plex has added support for this in the latest betas: https://support.plex.tv/articles/hdr-to-sdr-tone-mapping/

**Comment by LukePulverenti** [comment]
HDR Tone mapping when transcoding will be in Emby Server 4.6. Stay tuned to the beta channel over the next week if you'd like to help test.

**Comment by LukePulverenti** [comment]
https://emby.media/community/index.php?/topic/95213-46025-transcoding-tone-mapping/
